internal/executor/service/pod_issue_handler.go

```diff
@@ -0,0 +1,348 @@
+	issue.Reported = true
+}
+
+func (p *PodIssueService) HandlePodIssues() {
+	managedPods, err := p.clusterContext.GetBatchPods()
+	if err != nil {
+		log.WithError(err).Errorf("unable to handle pod issus as failed to load pods")
+	}
+	p.detectPodIssues(managedPods)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*2)
+	defer cancel()
+	p.handleKnownPodIssues(ctx, managedPods)
+}
+
+func (p *PodIssueService) detectPodIssues(allManagedPods []*v1.Pod) {
+	for _, pod := range allManagedPods {
+		if pod.DeletionTimestamp != nil && pod.DeletionTimestamp.Add(p.stuckTerminatingPodExpiry).Before(time.Now()) {
+			// pod is stuck in terminating phase, this sometimes happen on node failure
+			// it is safer to produce failed event than retrying as the job might have run already
+			issue := &podIssue{
+				OriginalPodState: pod.DeepCopy(),
+				JobId:            util.ExtractJobId(pod),
+				RunId:            util.ExtractJobRunId(pod),
+				Message:          "pod stuck in terminating phase, this might be due to platform problems",
+				Retryable:        false,
+				Type:             StuckTerminating,
+			}
+
+			p.registerIssue(issue)
+			break
+		} else if pod.Status.Phase == v1.PodUnknown || pod.Status.Phase == v1.PodPending {
+
+			podEvents, err := p.clusterContext.GetPodEvents(pod)
+			if err != nil {
+				log.Errorf("Unable to get pod events for pod %s: %v", pod.Name, err)
+			}
+
+			lastStateChange, err := util.LastStatusChange(pod)
+			if err != nil {
+				log.Errorf("Unable to get lastStateChange for pod %s: %v", pod.Name, err)
+				continue
+			}
+
+			action, cause, podCheckMessage := p.pendingPodChecker.GetAction(pod, podEvents, time.Now().Sub(lastStateChange))
+
+			if action != podchecks.ActionWait {
+				retryable := action == podchecks.ActionRetry
+				message := createStuckPodMessage(retryable, podCheckMessage)
+				podIssueType := StuckStartingUp
+				if cause == podchecks.NoNodeAssigned {
+					podIssueType = UnableToSchedule
+				}
+
+				log.Warnf("Found issue with pod %s in namespace %s: %s", pod.Name, pod.Namespace, message)
+
+				issue := &podIssue{
+					OriginalPodState: pod.DeepCopy(),
+					JobId:            util.ExtractJobId(pod),
+					RunId:            util.ExtractJobRunId(pod),
+					Message:          message,
+					Retryable:        retryable,
+					Type:             podIssueType,
+				}
+				p.registerIssue(issue)
+				break
+			}
+		}
+	}
+}
+
+func (p *PodIssueService) handleKnownPodIssues(ctx context.Context, allManagedPods []*v1.Pod) {
+	issues := createIssues(allManagedPods, p.knownPodIssues)
+	// Make issues from pods + issues
+	util.ProcessItemsWithThreadPool(ctx, 20, issues, p.handlePodIssue)
+}
+
+func createIssues(managedPods []*v1.Pod, podIssues map[string]*podIssue) []*issue {
+	podsByRunId := make(map[string]*v1.Pod, len(managedPods))
+
+	for _, pod := range managedPods {
+		runId := util.ExtractJobRunId(pod)
+		if runId != "" {
+			podsByRunId[runId] = pod
+		}
```

Comment:
what does it mean if `runId` is not present here?  Should we log something?

Reply:
Done